pulsar-broker/src/test/java/org/apache/pulsar/client/api/ConsumerRedeliveryTest.java

```diff
@@ -65,6 +68,19 @@ public Object[][] ackReceiptEnabled() {
         return new Object[][] { { true }, { false } };
     }
 
+    @DataProvider(name = "batchedMessageAck")
+    public Object[][] batchedMessageAck() {
+        // When batch index ack is disabled (by default), only after all single messages were sent would the pending
+        // ACK be added into the ACK tracker.
+        return new Object[][] {
+                // numAcked, batchSize, ack type
+                { 3, 5, CommandAck.AckType.Individual },
+                { 5, 5, CommandAck.AckType.Individual },
+                { 3, 5, CommandAck.AckType.Cumulative },
+                { 5, 5, CommandAck.AckType.Cumulative }
```

Comment:
5, 5 test case may not stable

Reply:
I see, I will take a look soon.

Reply:
It's fixed now. The root cause is that after https://github.com/apache/pulsar/pull/18486/commits/b0fad4076f5532103e9f688149048d7889376213, a BatchedMessageIdImpl will be compared with a MessageIdImpl, which should be forbidden.

Reply:
@congbobo184 PTAL again.